src/client/main.ts

```diff
@@ -420,24 +420,53 @@ function deserializeChannelChangedEvent(eventTransport: Transport<ChannelChanged
 
 if (typeof fin !== 'undefined') {
     getServicePromise().then((channelClient) => {
-        channelClient.register(APIToClientTopic.RECEIVE_INTENT, (payload: RaiseIntentPayload) => {
-            intentListeners.forEach((listener: IntentListener) => {
-                if (payload.intent === listener.intent) {
-                    listener.handler(payload.context);
+        channelClient.register(APIToClientTopic.RECEIVE_INTENT, async (payload: RaiseIntentPayload) => {
+            let successes = 0;
+            let failures = 0;
+
+            await Promise.all(intentListeners.filter((listener) => payload.intent === listener.intent).map(async (listener) => {
+                try {
+                    await listener.handler(payload.context);
+                    successes++;
+                } catch (e) {
+                    failures++;
+                    console.warn(`Error thrown by ${payload.intent} intent handler, swallowing error. Error message: ${e.message}`);
                 }
-            });
+            }));
+
+            if (failures > 0 && successes === 0) {
+                throw new Error(`All ${payload.intent} intent handlers failed`);
+            }
         });
 
         channelClient.register(APIToClientTopic.RECEIVE_CONTEXT, (payload: ReceiveContextPayload) => {
+            let successes = 0;
+            let failures = 0;
+
             contextListeners.forEach((listener: ContextListener) => {
-                listener.handler(payload.context);
+                try {
+                    listener.handler(payload.context);
+                    successes++;
+                } catch (e) {
+                    failures++;
+                    console.warn(`Error thrown by context handler, swallowing error. Error message: ${e.message}`);
+                }
             });
+
+            if (failures > 0 && successes === 0) {
+                throw new Error('All context handlers failed');
+            }
         });
 
         const eventHandler = getEventRouter();
 
         channelClient.register('event', (eventTransport: Targeted<Transport<Events>>) => {
-            eventHandler.dispatchEvent(eventTransport);
+            try {
+                eventHandler.dispatchEvent(eventTransport);
+            } catch (e) {
+                console.warn(`Error thrown dispatching ${eventTransport.type} event, rethrowing error. Error message: ${e.message}`);
+                throw e;
```

Comment:
General point, but sticking it here.

When re-throwing an error like this, does it preserve the initial stack trace? IIRC, not preserving the stack trace is of those special JavaScript "features"...

If it's overwritten, we should likely trace the initial stack trace as well as the errors (maybe in a collapsible way - `console.group` and stuff).

Reply:
This is the only place we're rethrowing in this PR, but I think it's kinda irrelevant to your point - in this case it gets thrown within the API handler, so bubbles up to the provider, where it then gets captured and logged. It's not possible for a client developer to get direct access to this errorobject anyway

What might be worthwhile is always logging the stack trace on the client side? But I think that's a separate question from whether stack traces are overwritten on re-throw

Reply:
Didn't realise this was in the client - true, a bit more specific than I was thinking.

Makes sense as-is.